From 1a92f7b8126c67b88d91513e6d731a61579f8ef7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Sebastian=20Dr=C3=B6ge?= <sebastian.droege@collabora.co.uk>
Date: Thu, 9 Aug 2012 10:25:38 +0200
Subject: [PATCH 1/2] configure: Fix GTK required version variable name

---
 configure.ac |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 2a16a18..848c92f 100644
--- a/configure.ac
+++ b/configure.ac
@@ -234,7 +234,7 @@ AC_SUBST(GSTPB_PREFIX)
 
 dnl GTK is optional and used in examples
 HAVE_GTK=no
-GT_REQ=3.0.0
+GTK_REQ=3.0.0
 if test "x$BUILD_EXAMPLES" = "xyes"; then
   PKG_CHECK_MODULES(GTK, gtk+-$with_gtk >= $GTK_REQ, HAVE_GTK=yes, HAVE_GTK=no)
   dnl some examples need gtk+-x11
-- 
1.7.10.4

