From: Jim Hodapp <jim.hodapp@canonical.com>
Date: Tue, 14 Oct 2014 14:56:57 -0400
Subject: [PATCH] Handle more rotation transformation matrices

Origin: vendor
Forwarded: no

Index: b/gst/isomp4/qtdemux.c
===================================================================
--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -7634,6 +7634,17 @@
                                        (m)[3] == (d << 16) && (m)[4] == (e << 16) && \
                                        (m)[6] == (g << 16) && (m)[7] == (h << 16))
 
+  const float inv2exp16 = 1.0f / (float) (1 << 16);
+  const float inv2exp30 = 1.0f / (float) (1 << 30);
+
+  /* Print out the normalized transformation matrix */
+  GST_DEBUG_OBJECT (qtdemux, "%.2f %.2f %.2f", ((int) (matrix[0]) * inv2exp16),
+      ((int) (matrix[1]) * inv2exp16), ((int) (matrix[2]) * inv2exp30));
+  GST_DEBUG_OBJECT (qtdemux, "%.2f %.2f %.2f", ((int) (matrix[3]) * inv2exp16),
+      ((int) (matrix[4]) * inv2exp16), ((int) (matrix[5]) * inv2exp30));
+  GST_DEBUG_OBJECT (qtdemux, "%.2f %.2f %.2f", ((int) (matrix[6]) * inv2exp16),
+      ((int) (matrix[7]) * inv2exp16), ((int) (matrix[8]) * inv2exp30));
+
   /* only handle the cases where the last column has standard values */
   if (matrix[2] == 0 && matrix[5] == 0 && matrix[8] == 1 << 30) {
     const gchar *rotation_tag = NULL;
@@ -7642,13 +7653,16 @@
     if (QTCHECK_MATRIX (matrix, 1, 0, 0, 1, 0, 0)) {
       /* NOP */
     } else if (QTCHECK_MATRIX (matrix, 0, 1, G_MAXUINT16, 0,
-            stream->display_height, 0)) {
+            stream->display_height, 0) ||
+        QTCHECK_MATRIX (matrix, 0, 1, G_MAXUINT16, 0, 0, 0)) {
       rotation_tag = "rotate-90";
     } else if (QTCHECK_MATRIX (matrix, G_MAXUINT16, 0, 0, G_MAXUINT16,
-            stream->display_width, stream->display_height)) {
+            stream->display_width, stream->display_height) ||
+        QTCHECK_MATRIX (matrix, G_MAXUINT16, 0, 0, G_MAXUINT16, 0, 0)) {
       rotation_tag = "rotate-180";
     } else if (QTCHECK_MATRIX (matrix, 0, G_MAXUINT16, 1, 0, 0,
-            stream->display_width)) {
+            stream->display_width) ||
+        QTCHECK_MATRIX (matrix, 0, G_MAXUINT16, 1, 0, 0, 0)) {
       rotation_tag = "rotate-270";
     } else {
       GST_FIXME_OBJECT (qtdemux, "Unhandled transformation matrix values");
